jina/types/arrays/document.py

```diff
@@ -98,24 +107,48 @@ class DocumentArray(
     It gives an efficient view of a list of Document. One can iterate over it like
     a generator but ALSO modify it, count it, get item, or union two 'DocumentArray's using the '+' and '+=' operators.
 
-    :param doc_views: A list of :class:`Document`
-    :type doc_views: Optional[Union['RepeatedContainer', Iterable['Document']]]
+    It is supposed to act as a view containing a pointer to a `RepeatedContainer` of `DocumentProto` while offering `Document` Jina native types
+    when getting items or iterating over it
+
+    :param docs: the document array to construct from. One can also give `DocumentArrayProto` directly, then depending on the ``copy``,
+                it builds a view or a copy from it. It also can accept a List
+    :param copy: when ``docs`` is given as a :class:`DocumentArrayProto` object, build a
+                view (i.e. weak reference) from it or a deep copy from it.
     """
 
-    def __init__(
-        self,
-        doc_views: Optional[Union['RepeatedContainer', Iterable['Document']]] = None,
-    ):
+    def __init__(self, docs: Optional[DocumentArraySourceType] = None):
         super().__init__()
-        if doc_views is not None:
-            from .memmap import DocumentArrayMemmap
-
-            if isinstance(doc_views, (Generator, DocumentArrayMemmap)):
-                self._doc_views = list(doc_views)
+        self._pb_body = []
+        if docs is not None:
+            if isinstance(docs, jina_pb2.DocumentArrayProto):
+                # This would happen when loading from file or memmap
+                self._pb_body = docs.docs
+            elif isinstance(docs, RepeatedContainer):
+                # This would happen when `doc.matches` or `doc.chunks`
+                self._pb_body = docs
+            elif isinstance(docs, DocumentArray):
+                # This would happen in the client
+                self._pb_body = docs._pb_body
+            elif isinstance(docs, list):
+                # This would happen in the client
+                for doc in docs:
+                    if isinstance(doc, Document):
+                        self._pb_body.append(doc.proto)
+                    elif isinstance(doc, jina_pb2.DocumentProto):
+                        self._pb_body.append(doc)
+                    else:
+                        raise ValueError(f'Unexpected element in an input list')
```

Comment:
so this introduces extra-copying, we should then reduce the overall usage of this constructor